File: Schimpf 2024 connected reformulation counterexample/src/exact.js
/**
 * Exact arithmetic for rational functions in p with coefficients in Q.
 *
 * Polynomials are coefficient arrays in ascending order:
 *   [a0, a1, ...] = a0 + a1*p + ...
 */

export class Fraction {
  constructor(numerator, denominator = 1n) {
    let n = typeof numerator === "bigint" ? numerator : BigInt(numerator);
    let d = typeof denominator === "bigint" ? denominator : BigInt(denominator);
    if (d === 0n) throw new Error("zero denominator");
    if (d < 0n) {
      n = -n;
      d = -d;
    }
    const divisor = Fraction.gcd(n < 0n ? -n : n, d);
    this.n = n / divisor;
    this.d = d / divisor;
    Object.freeze(this);
  }

  static gcd(a, b) {
    while (b !== 0n) [a, b] = [b, a % b];
    return a || 1n;
  }

  static of(value) {
    return value instanceof Fraction ? value : new Fraction(value);
  }

  add(other) {
    const rhs = Fraction.of(other);
    return new Fraction(this.n * rhs.d + rhs.n * this.d, this.d * rhs.d);
  }

  sub(other) {
    return this.add(Fraction.of(other).neg());
  }

  mul(other) {
    const rhs = Fraction.of(other);
    return new Fraction(this.n * rhs.n, this.d * rhs.d);
  }

  div(other) {
    const rhs = Fraction.of(other);
    if (rhs.n === 0n) throw new Error("division by zero");
    return new Fraction(this.n * rhs.d, this.d * rhs.n);
  }

  neg() {
    return new Fraction(-this.n, this.d);
  }

  eq(other) {
    const rhs = Fraction.of(other);
    return this.n === rhs.n && this.d === rhs.d;
  }

  isZero() {
    return this.n === 0n;
  }

  toJSON() {
    return this.toString();
  }

  toString() {
    return this.d === 1n ? `${this.n}` : `${this.n}/${this.d}`;
  }
}

export const q = (numerator, denominator = 1) =>
  new Fraction(BigInt(numerator), BigInt(denominator));

export function polynomial(values) {
  return trimPolynomial(values.map(Fraction.of));
}

export function trimPolynomial(values) {
  let length = values.length;
  while (length > 1 && values[length - 1].isZero()) length -= 1;
  return values.slice(0, Math.max(1, length));
}

export function polynomialIsZero(values) {
  return trimPolynomial(values).every((coefficient) => coefficient.isZero());
}

export function addPolynomials(left, right) {
  const length = Math.max(left.length, right.length);
  return trimPolynomial(
    Array.from({ length }, (_, index) =>
      (left[index] || q(0)).add(right[index] || q(0))
    )
  );
}

export function negatePolynomial(values) {
  return values.map((coefficient) => coefficient.neg());
}

export function subtractPolynomials(left, right) {
  return addPolynomials(left, negatePolynomial(right));
}

export function scalePolynomial(values, scalar) {
  const factor = Fraction.of(scalar);
  return trimPolynomial(values.map((coefficient) => coefficient.mul(factor)));
}

export function multiplyPolynomials(left, right) {
  const result = Array.from(
    { length: left.length + right.length - 1 },
    () => q(0)
  );
  for (let i = 0; i < left.length; i += 1) {
    for (let j = 0; j < right.length; j += 1) {
      result[i + j] = result[i + j].add(left[i].mul(right[j]));
    }
  }
  return trimPolynomial(result);
}

export function dividePolynomials(dividend, divisor) {
  const denominator = trimPolynomial(divisor);
  if (polynomialIsZero(denominator)) throw new Error("polynomial division by zero");
  let remainder = trimPolynomial(dividend);
  const quotient = Array.from(
    { length: Math.max(1, remainder.length - denominator.length + 1) },
    () => q(0)
  );
  while (
    !polynomialIsZero(remainder) &&
    remainder.length >= denominator.length
  ) {
    const degree = remainder.length - denominator.length;
    const factor = remainder[remainder.length - 1].div(
      denominator[denominator.length - 1]
    );
    quotient[degree] = quotient[degree].add(factor);
    const shifted = Array.from({ length: degree }, () => q(0)).concat(
      scalePolynomial(denominator, factor)
    );
    remainder = subtractPolynomials(remainder, shifted);
  }
  return {
    quotient: trimPolynomial(quotient),
    remainder: trimPolynomial(remainder),
  };
}

export function polynomialGcd(left, right) {
  let a = trimPolynomial(left);
  let b = trimPolynomial(right);
  while (!polynomialIsZero(b)) {
    const { remainder } = dividePolynomials(a, b);
    a = b;
    b = remainder;
  }
  if (polynomialIsZero(a)) return polynomial([1]);
  return scalePolynomial(a, q(1).div(a[a.length - 1]));
}

export function evaluatePolynomial(values, point) {
  const x = Fraction.of(point);
  let result = q(0);
  for (let index = values.length - 1; index >= 0; index -= 1) {
    result = result.mul(x).add(values[index]);
  }
  return result;
}

export function rational(numerator, denominator = polynomial([1])) {
  let top = trimPolynomial(numerator);
  let bottom = trimPolynomial(denominator);
  if (polynomialIsZero(bottom)) throw new Error("zero rational-function denominator");
  if (polynomialIsZero(top)) {
    return Object.freeze({ numerator: polynomial([0]), denominator: polynomial([1]) });
  }
  const divisor = polynomialGcd(top, bottom);
  top = dividePolynomials(top, divisor).quotient;
  bottom = dividePolynomials(bottom, divisor).quotient;
  const normalization = bottom[0].isZero()
    ? bottom[bottom.length - 1]
    : bottom[0];
  top = scalePolynomial(top, q(1).div(normalization));
  bottom = scalePolynomial(bottom, q(1).div(normalization));
  return Object.freeze({ numerator: top, denominator: bottom });
}

export function addRationals(left, right) {
  return rational(
    addPolynomials(
      multiplyPolynomials(left.numerator, right.denominator),
      multiplyPolynomials(right.numerator, left.denominator)
    ),
    multiplyPolynomials(left.denominator, right.denominator)
  );
}

export function subtractRationals(left, right) {
  return rational(
    subtractPolynomials(
      multiplyPolynomials(left.numerator, right.denominator),
      multiplyPolynomials(right.numerator, left.denominator)
    ),
    multiplyPolynomials(left.denominator, right.denominator)
  );
}

export function multiplyRationals(left, right) {
  return rational(
    multiplyPolynomials(left.numerator, right.numerator),
    multiplyPolynomials(left.denominator, right.denominator)
  );
}

export function divideRationals(left, right) {
  if (polynomialIsZero(right.numerator)) throw new Error("rational division by zero");
  return rational(
    multiplyPolynomials(left.numerator, right.denominator),
    multiplyPolynomials(left.denominator, right.numerator)
  );
}

export function scaleRational(value, scalar) {
  return rational(scalePolynomial(value.numerator, scalar), value.denominator);
}

export function rationalEquals(left, right) {
  return polynomialIsZero(
    subtractPolynomials(
      multiplyPolynomials(left.numerator, right.denominator),
      multiplyPolynomials(right.numerator, left.denominator)
    )
  );
}

export function shiftedAtOne(values) {
  const result = Array.from({ length: values.length }, () => q(0));
  for (let degree = 0; degree < values.length; degree += 1) {
    let binomial = 1n;
    for (let index = 0; index <= degree; index += 1) {
      if (index > 0) {
        binomial =
          (binomial * BigInt(degree - index + 1)) / BigInt(index);
      }
      result[index] = result[index].add(
        values[degree].mul(new Fraction(binomial))
      );
    }
  }
  return trimPolynomial(result);
}

export function leadingPolynomialAtOne(values) {
  const shifted = shiftedAtOne(values);
  for (let order = 0; order < shifted.length; order += 1) {
    if (!shifted[order].isZero()) return { order, coefficient: shifted[order] };
  }
  return { order: Number.POSITIVE_INFINITY, coefficient: q(0) };
}

export function leadingRationalAtOne(value) {
  const top = leadingPolynomialAtOne(value.numerator);
  const bottom = leadingPolynomialAtOne(value.denominator);
  return {
    order: top.order - bottom.order,
    coefficient: top.coefficient.div(bottom.coefficient),
  };
}

export function laurentAtOne(value, minimumPower, maximumPower) {
  const shiftedNumerator = shiftedAtOne(value.numerator);
  const shiftedDenominator = shiftedAtOne(value.denominator);
  const numeratorLead = leadingPolynomialAtOne(value.numerator);
  const denominatorLead = leadingPolynomialAtOne(value.denominator);
  if (numeratorLead.order === Number.POSITIVE_INFINITY) {
    return new Map(
      Array.from(
        { length: maximumPower - minimumPower + 1 },
        (_, index) => [minimumPower + index, q(0)]
      )
    );
  }
  const start = numeratorLead.order - denominatorLead.order;
  const numerator = shiftedNumerator.slice(numeratorLead.order);
  const denominator = shiftedDenominator.slice(denominatorLead.order);
  const needed = Math.max(0, maximumPower - start);
  const coefficients = [];
  for (let index = 0; index <= needed; index += 1) {
    let rhs = numerator[index] || q(0);
    for (let offset = 1; offset <= index; offset += 1) {
      rhs = rhs.sub(
        (denominator[offset] || q(0)).mul(coefficients[index - offset])
      );
    }
    coefficients[index] = rhs.div(denominator[0]);
  }
  const output = new Map();
  for (let power = minimumPower; power <= maximumPower; power += 1) {
    const index = power - start;
    output.set(
      power,
      index >= 0 && index < coefficients.length ? coefficients[index] : q(0)
    );
  }
  return output;
}

export function polynomialToString(values, variable = "p") {
  const terms = [];
  for (let degree = 0; degree < values.length; degree += 1) {
    const coefficient = values[degree];
    if (coefficient.isZero()) continue;
    const negative = coefficient.n < 0n;
    const magnitude = new Fraction(
      negative ? -coefficient.n : coefficient.n,
      coefficient.d
    ).toString();
    const variablePart =
      degree === 1 ? variable : `${variable}^${degree}`;
    const body =
      degree === 0
        ? magnitude
        : magnitude === "1"
          ? variablePart
          : `${magnitude}*${variablePart}`;
    terms.push({ sign: negative ? "-" : "+", body });
  }
  if (terms.length === 0) return "0";
  const [first, ...rest] = terms;
  let output = first.sign === "-" ? `-${first.body}` : first.body;
  for (const term of rest) output += ` ${term.sign} ${term.body}`;
  return output;
}

export function rationalToString(value, variable = "p") {
  const numerator = polynomialToString(value.numerator, variable);
  const denominator = polynomialToString(value.denominator, variable);
  return denominator === "1" ? numerator : `(${numerator})/(${denominator})`;
}

export function rationalToJSON(value) {
  return {
    numerator: value.numerator.map(String),
    denominator: value.denominator.map(String),
    expression: rationalToString(value),
  };
}

File: Schimpf 2024 connected reformulation counterexample/schimpf_counterexample/exact.py
"""Exact arithmetic for rational functions in one variable.

Polynomials are tuples of ``fractions.Fraction`` coefficients in ascending
order.  Thus ``(a_0, a_1, ...)`` represents ``a_0 + a_1*p + ...``.
"""

from __future__ import annotations

from dataclasses import dataclass
from fractions import Fraction
from math import comb
from typing import Iterable

Polynomial = tuple[Fraction, ...]


def polynomial(values: Iterable[int | Fraction]) -> Polynomial:
    return trim(tuple(Fraction(value) for value in values))


def trim(values: Polynomial) -> Polynomial:
    if not values:
        return (Fraction(0),)
    length = len(values)
    while length > 1 and values[length - 1] == 0:
        length -= 1
    return values[:length]


ZERO_POLYNOMIAL = polynomial([0])
ONE_POLYNOMIAL = polynomial([1])


def is_zero(values: Polynomial) -> bool:
    return trim(values) == ZERO_POLYNOMIAL


def add_polynomials(left: Polynomial, right: Polynomial) -> Polynomial:
    length = max(len(left), len(right))
    return trim(
        tuple(
            (left[index] if index < len(left) else Fraction(0))
            + (right[index] if index < len(right) else Fraction(0))
            for index in range(length)
        )
    )


def negate_polynomial(values: Polynomial) -> Polynomial:
    return tuple(-coefficient for coefficient in values)


def subtract_polynomials(left: Polynomial, right: Polynomial) -> Polynomial:
    return add_polynomials(left, negate_polynomial(right))


def scale_polynomial(
    values: Polynomial, scalar: int | Fraction
) -> Polynomial:
    factor = Fraction(scalar)
    return trim(tuple(coefficient * factor for coefficient in values))


def multiply_polynomials(left: Polynomial, right: Polynomial) -> Polynomial:
    result = [Fraction(0)] * (len(left) + len(right) - 1)
    for left_degree, left_coefficient in enumerate(left):
        for right_degree, right_coefficient in enumerate(right):
            result[left_degree + right_degree] += (
                left_coefficient * right_coefficient
            )
    return trim(tuple(result))


def divide_polynomials(
    dividend: Polynomial, divisor: Polynomial
) -> tuple[Polynomial, Polynomial]:
    denominator = trim(divisor)
    if is_zero(denominator):
        raise ZeroDivisionError("polynomial division by zero")
    remainder = trim(dividend)
    quotient = [Fraction(0)] * max(
        1, len(remainder) - len(denominator) + 1
    )
    while not is_zero(remainder) and len(remainder) >= len(denominator):
        degree = len(remainder) - len(denominator)
        factor = remainder[-1] / denominator[-1]
        quotient[degree] += factor
        shifted = (Fraction(0),) * degree + scale_polynomial(
            denominator, factor
        )
        remainder = subtract_polynomials(remainder, shifted)
    return trim(tuple(quotient)), trim(remainder)


def polynomial_gcd(left: Polynomial, right: Polynomial) -> Polynomial:
    a = trim(left)
    b = trim(right)
    while not is_zero(b):
        _, remainder = divide_polynomials(a, b)
        a, b = b, remainder
    if is_zero(a):
        return ONE_POLYNOMIAL
    return scale_polynomial(a, Fraction(1, 1) / a[-1])


def evaluate_polynomial(
    values: Polynomial, point: int | Fraction
) -> Fraction:
    x = Fraction(point)
    result = Fraction(0)
    for coefficient in reversed(values):
        result = result * x + coefficient
    return result


@dataclass(frozen=True, init=False)
class RationalFunction:
    """A reduced rational function over the rational numbers."""

    numerator: Polynomial
    denominator: Polynomial

    def __init__(
        self,
        numerator: Iterable[int | Fraction],
        denominator: Iterable[int | Fraction] = (1,),
    ) -> None:
        top = polynomial(numerator)
        bottom = polynomial(denominator)
        if is_zero(bottom):
            raise ZeroDivisionError("zero rational-function denominator")
        if is_zero(top):
            object.__setattr__(self, "numerator", ZERO_POLYNOMIAL)
            object.__setattr__(self, "denominator", ONE_POLYNOMIAL)
            return

        divisor = polynomial_gcd(top, bottom)
        top, top_remainder = divide_polynomials(top, divisor)
        bottom, bottom_remainder = divide_polynomials(bottom, divisor)
        if not is_zero(top_remainder) or not is_zero(bottom_remainder):
            raise ArithmeticError("polynomial gcd did not divide exactly")

        normalization = bottom[0] if bottom[0] != 0 else bottom[-1]
        top = scale_polynomial(top, Fraction(1, 1) / normalization)
        bottom = scale_polynomial(bottom, Fraction(1, 1) / normalization)
        object.__setattr__(self, "numerator", top)
        object.__setattr__(self, "denominator", bottom)

    @staticmethod
    def coerce(value: RationalFunction | int | Fraction) -> RationalFunction:
        if isinstance(value, RationalFunction):
            return value
        return RationalFunction([Fraction(value)])

    def __add__(
        self, other: RationalFunction | int | Fraction
    ) -> RationalFunction:
        right = RationalFunction.coerce(other)
        return RationalFunction(
            add_polynomials(
                multiply_polynomials(self.numerator, right.denominator),
                multiply_polynomials(right.numerator, self.denominator),
            ),
            multiply_polynomials(self.denominator, right.denominator),
        )

    def __radd__(
        self, other: RationalFunction | int | Fraction
    ) -> RationalFunction:
        return self + other

    def __neg__(self) -> RationalFunction:
        return RationalFunction(
            negate_polynomial(self.numerator), self.denominator
        )

    def __sub__(
        self, other: RationalFunction | int | Fraction
    ) -> RationalFunction:
        return self + (-RationalFunction.coerce(other))

    def __rsub__(
        self, other: RationalFunction | int | Fraction
    ) -> RationalFunction:
        return RationalFunction.coerce(other) - self

    def __mul__(
        self, other: RationalFunction | int | Fraction
    ) -> RationalFunction:
        right = RationalFunction.coerce(other)
        return RationalFunction(
            multiply_polynomials(self.numerator, right.numerator),
            multiply_polynomials(self.denominator, right.denominator),
        )

    def __rmul__(
        self, other: RationalFunction | int | Fraction
    ) -> RationalFunction:
        return self * other

    def __truediv__(
        self, other: RationalFunction | int | Fraction
    ) -> RationalFunction:
        right = RationalFunction.coerce(other)
        if is_zero(right.numerator):
            raise ZeroDivisionError("rational-function division by zero")
        return RationalFunction(
            multiply_polynomials(self.numerator, right.denominator),
            multiply_polynomials(self.denominator, right.numerator),
        )

    def __rtruediv__(
        self, other: RationalFunction | int | Fraction
    ) -> RationalFunction:
        return RationalFunction.coerce(other) / self

    def is_zero(self) -> bool:
        return is_zero(self.numerator)

    def to_dict(self) -> dict[str, object]:
        return {
            "numerator": [str(value) for value in self.numerator],
            "denominator": [str(value) for value in self.denominator],
            "expression": str(self),
        }

    def __str__(self) -> str:
        top = polynomial_to_string(self.numerator)
        bottom = polynomial_to_string(self.denominator)
        return top if bottom == "1" else f"({top})/({bottom})"


ZERO = RationalFunction([0])
ONE = RationalFunction([1])


def shifted_at_one(values: Polynomial) -> Polynomial:
    result = [Fraction(0)] * len(values)
    for degree, coefficient in enumerate(values):
        for index in range(degree + 1):
            result[index] += coefficient * comb(degree, index)
    return trim(tuple(result))


def leading_polynomial_at_one(
    values: Polynomial,
) -> tuple[int | None, Fraction]:
    shifted = shifted_at_one(values)
    for order, coefficient in enumerate(shifted):
        if coefficient != 0:
            return order, coefficient
    return None, Fraction(0)


def leading_rational_at_one(
    value: RationalFunction,
) -> tuple[int | None, Fraction]:
    top_order, top_coefficient = leading_polynomial_at_one(value.numerator)
    bottom_order, bottom_coefficient = leading_polynomial_at_one(
        value.denominator
    )
    if top_order is None:
        return None, Fraction(0)
    if bottom_order is None:
        raise ZeroDivisionError("identically zero denominator")
    return top_order - bottom_order, top_coefficient / bottom_coefficient


def laurent_at_one(
    value: RationalFunction, minimum_power: int, maximum_power: int
) -> dict[int, Fraction]:
    top_order, _ = leading_polynomial_at_one(value.numerator)
    bottom_order, _ = leading_polynomial_at_one(value.denominator)
    if top_order is None:
        return {
            power: Fraction(0)
            for power in range(minimum_power, maximum_power + 1)
        }
    if bottom_order is None:
        raise ZeroDivisionError("identically zero denominator")

    numerator = shifted_at_one(value.numerator)[top_order:]
    denominator = shifted_at_one(value.denominator)[bottom_order:]
    start = top_order - bottom_order
    needed = max(0, maximum_power - start)
    coefficients: list[Fraction] = []
    for index in range(needed + 1):
        right = (
            numerator[index] if index < len(numerator) else Fraction(0)
        )
        for offset in range(1, index + 1):
            denominator_coefficient = (
                denominator[offset]
                if offset < len(denominator)
                else Fraction(0)
            )
            right -= denominator_coefficient * coefficients[index - offset]
        coefficients.append(right / denominator[0])

    output: dict[int, Fraction] = {}
    for power in range(minimum_power, maximum_power + 1):
        index = power - start
        output[power] = (
            coefficients[index]
            if 0 <= index < len(coefficients)
            else Fraction(0)
        )
    return output


def polynomial_to_string(values: Polynomial, variable: str = "p") -> str:
    terms: list[tuple[str, str]] = []
    for degree, coefficient in enumerate(values):
        if coefficient == 0:
            continue
        sign = "-" if coefficient < 0 else "+"
        magnitude = abs(coefficient)
        if degree == 0:
            body = str(magnitude)
        else:
            variable_part = variable if degree == 1 else f"{variable}^{degree}"
            body = (
                variable_part
                if magnitude == 1
                else f"{magnitude}*{variable_part}"
            )
        terms.append((sign, body))
    if not terms:
        return "0"
    first_sign, first_body = terms[0]
    output = f"-{first_body}" if first_sign == "-" else first_body
    for sign, body in terms[1:]:
        output += f" {sign} {body}"
    return output
